### ISSUE TITLE
Nonce increment when transaction failed

### ISSUE THREAD

**Issue by gituser** [paritytech/parity-ethereum#9040]
> I'm running:
>
> - **Which Parity version?**: 1.10.8
> - **Which operating system?**: Linux
> - **How installed?**: from source
> - **Are you fully synchronized?**: yes
> - **Which network are you connected to?**: ethereum 
> - **Did you try to restart the node?**: yes

**Summary**
When you try to send a transaction with gas less than 21000 or (if you specify amount of the transaction bigger than your actual address balance) it fails, but nonce still increments, thus all future transactions will be forever in the queue, until someone fills the nonce gap between that last successful transaction and the one in the queue or restarts the parity.

**Here is how to reproduce:**
1) create some transaction without nonce (thus nonce will be picked up, let's say `nonce` is 10)
2) next, create a transaction with gas=20000, you'll get this error:
```
Error: Transaction gas is too low. There is not enough gas to cover minimal cost of the transaction (minimal: 21000, got: 20000). Try increasing supplied gas.
```
3) now try creating another transaction without specifying the nonce, `nonce` will be 12 and thus transaction will stay in the pool, until you send transaction with `nonce=11`

I think nonce shouldn't be incremented if transaction has failed due to low gas or any other error when there is no resulting hash.

**Comment by gituser** [comment]
The issue here is not about gas amount, but rather about nonce increment. Why parity does increment the nonce when in fact there was no previous transaction?

**Comment by gituser** [comment]
Also, `nonce` is being incremented if you try to send transaction which exceeds your current account balance. That means that after that all future transactions will be stuck as well.

**Comment by seunlanlege** [comment]
can't replicate this on kovan.

**Comment by Tbaut** [comment]
@gituser please provide a concrete example (with rpc or a web3 script) that could allow us to reproduce the behavior? Happy to reopen then.

**Comment by gituser** [comment]
You want me to provide you with RPC call how to send transaction? It's all described in the issue description, re-read it.

Simply try to send transaction with gas=20000 or specify value bigger than your current balance and then try to create another transaction with good value, it will never be picked up, because there will be a gap in nonce.

Why close the issue without trying to reproduce it? I've written exact steps how to reproduce the issue.

**Comment by gituser** [comment]
Also, I've just tested on newest parity release and the bug is still there on Ropsten, I'm pretty sure on Kovan it should be there too.

**Comment by Tbaut** [comment]
The reason I ask is not to bother you :) So user introduce mistake in their RPC and call it a bug. Since @seunlanlege gave it a try and couldn't reproduce, that's a good enough reason to ask you some more info ;) Thanks for providing them, we'll check on Ropsten.

**Comment by gituser** [comment]
Sure, it's non-standard behavior, but software should be able to handle such use-cases and not break, don't you think?
And for example not everyone know that minimum gas amount for transaction is 21000.

**Comment by seunlanlege** [comment]
closed via #10051
